packages/dev/core/src/FlowGraph/flowGraphConnectionPoint.ts

```diff
@@ -0,0 +1,76 @@
+import type { Nullable } from "../types";
+import type { FlowGraphBlock } from "./flowGraphBlock";
+import type { FlowGraphExecutionBlock } from "./flowGraphExecutionBlock";
+import { isDataUpdater } from "./iDataUpdater";
+
+export enum FlowGraphConnectionPointDirection {
+    Input,
+    Output,
+}
+
+/**
+ * @experimental
+ * Represents a connection point for a signal.
+ * When an output point is activated, it will activate the connected input point.
+ * When an input point is activated, it will execute the block it belongs to.
+ */
+export class FlowGraphSignalConnectionPoint {
+    name: string;
+    direction: FlowGraphConnectionPointDirection;
+    ownerBlock: FlowGraphExecutionBlock;
+    connectedPoint: Nullable<FlowGraphSignalConnectionPoint>;
+
+    connectTo(point: FlowGraphSignalConnectionPoint): void {
+        if (this.direction === point.direction) {
+            throw new Error("Cannot connect two points of the same direction");
+        }
+        this.connectedPoint = point;
+        point.connectedPoint = this;
+    }
+
+    activateSignal(): void {
+        if (this.direction === FlowGraphConnectionPointDirection.Input) {
+            this.ownerBlock.execute();
+        } else {
+            this.connectedPoint?.activateSignal();
+        }
+    }
+}
+
+/**
+ * @experimental
+ * Represents a connection point for data.
+ * An unconnected input point can have a default value.
+ * An output point will only have a value if it is connected to an input point. Furthermore,
+ * if the point belongs to a "function" node, the node will run its function to update the value.
+ */
+export class FlowGraphDataConnectionPoint<T> {

```

Comment:
The problem I had when trying to unify the classes was the connectedPoint's type - I wanted a SignalConnectionPoint to only connect to another SignalConnectionPoint - and the same for DataConnectionPoint. Through, I can always check that on the connectTo function, like I'm doing with the direction...

Reply:
Oh, also, ownerBlock's typing was also a point: I wanted a SignalConnectionPoint to only have a ownerBlock of type ExecutionBlock, because the function of the SignalConnectionPoint is to call _execute from the ExecutionBlock. Through the same from my previous comments applies - I could just check for typings at runtime too...?

Reply:
I've tried something in the Typescript sandbox (I've simplified as much as possible to retain only the elements useful to the discussion):

[Link to snippet](https://www.typescriptlang.org/play?#code/MYGwhgzhAEBiIHsDuBxATmADgCwEKOAGtoBvAXwCgLRIZ5l0tsBRADwFNgBXAFwEsEAO3wIi0dqx7tBAEzqJUGHCLHkqYAEYQeGYD2g0ocBYxwBhIYM78hABQR9B+khWhvomNAil72M6AD6wJbWfvaOPABcxgxK2BaCVnoCguFO0AA+0IJcICDQALzZuSAA3FTuBkLaaFx6CGgAFJ7eof4ByFZoKoTR9IpMPQCUpJSu7sGJ1gAqCM0OTn0mcQlJNqkLPCMulZU82HwQAHRBIb4yafpFmJvlu+43ESeTa2GbhdD7h3fulGOG8liTAAImAeGBVtYUpdxJJpHIYgNzGd1jCdhNqjo6jwGs0vD4pO1OuxugReojTHgydtxvcIFxMCTGh0kF1hj83GNKi8ZnNHosKXFQeDIck7Jsafc3PTGWgjjy9LN5hEhhzoH8qADBUwAMp8ADmgjAIFFqPeEikskBSPiKOh73RbkmNWxuJaBL8gWJpNE5P6lLYnF4KWGpFpuxlTJZbOpaq5GKmir5myWQJwesNxtN9pVYal0EjcoVPCV-K2cYo-0x0A0ZI+ViQ2uUZMaquo1YkQakPXr7Eb-rige46x6rfK7cE2mgMjBELt4oivf7yxBs+zC6cjVrvrbzv0EANRpN842i6KDab2AzR-Xp8326IqugAHpnwW6sB2H4YAUAHzQAA5AB5aZoCAgBpABCCcpwPTNj0TM0IgAJiXS9ryzE9LkaTth3YdkqBnEUsM2eUUSVIi50QnMnCfV932AT9v0Kf8EEIChKNvS4yMQpU4JvEiVVKF83wAMzAPgQB-Vj2IofjMOojceB4tYKLXQTaOE+jxMk6ToDYuTDwUtYaOU4s+KMhCTKU5C6LfelGK-BE-309igA)

But I'm not a Typescript expert and I can't get the `FlowGraphSignalConnectionPoint` constructor to accept only an instance of `FlowGraphExecutionBlock` (whereas it works for `connectTo` which accepts only an instance of `FlowGraphSignalConnectionPoint` and not a `FlowGraphConnectionPoint`) => see line 40

Let's ask @RaananW !

Reply:
Great idea of using the sandbox! Makes the problem a bit easier to think about. What if we make the base class' constructor empty and the subclasses' constructors with the expected types, [like this](https://www.typescriptlang.org/play?#code/MYGwhgzhAEBiIHsDuBxATmADgCwEKOAGtoBvAXwCgLRIZ5l0tsBRADwFNgBXAFwEsEAO3wIi0dqx7tBAEzqJUGHCLEkK0DeI7cpACgCUpdZpPAhEBCHYA6RAHNdAchkJoFgLbse2PoLuP9AG4TYw1KSgowACMIHgxgHmgaKDgFRhwAYSFBTn4hAAUEX0S1E0w0BCkE9hloAH0zQRzqmULigC5UhiVsLKbcgUE2wUSAH2hBLhAQaABeCamQQNDocsrcmvrkHLQVQk76RSY96HHJ6bmF6eWVxti0LgSENANSCNNs3IAVBF1MIpGBzSPT6zTyQwBPEMpRMmm8fAg1ganxaw0S83+xWWsM0mJGSMaYJqaMu8Ig2M0700EC8AHkkDtdNt2Gggd1jgRCNCViYyUjmbtOZcBRSwhQIsl5OycAARMA8MCggYFSFaKSyKVHTIo8EkmEaO5xR48Z66OoCvZsrV4TncnEaCBcTAsgyi2F8mk8emM80Mll7IIrKkGnU-P6Qq3pbByhVKhKDNF2+2O51oayE76-PFQt3hKiSrrWgDKfDsgjAIDjutVEnVckLUarCdV+qS5iNTxevp2lobPTYnF4gwDRntbidLsDY49dL9XYtttzVA+-QSYezkZ6JbLFabKuKSZxKZZ6dDWchU8p4vz7egUSF8xySD7HNEhFd1FvEkHUhOj-Yz6HFGA46MOnIfp+gixNAMjyoqOrNsUlxPi+spwXuELFLo95voGhpuKW5aVgh+4jMhAGodg25ERhaLYYu0AAPSMdAABmYB8CAMCzAAfNACCEAAhJB0EQIRu4kZhIwAEzkYBwJMNREmrtWWHfjo7ABjcsGxpJaKnipYY6fBKmISMQRMSxjrAMA7A1NxfECRQxm0ZCBlgmGYk7sRpmkTmllsRxXFzI5hBUF5NF6W5GZrr8LlRQewTMYFnEOfxYURcpYJmTw7mZromU+dlfnSRZyXWbZ9khelFBAA)? It's failing for the constructor case now. The only caveat is having to use ! for _ownerBlock.

Reply:
Good idea!

Here's [a new one](https://www.typescriptlang.org/play?#code/MYGwhgzhAEBiIHsDuBxATmADgCwEKOAGtoBvAXwCgLRIZ5l0tsBRADwFNgBXAFwEsEAO3wIi0dqx7tBAEzqJUGHCLEkK0DeI7cpACgCUpdZpPAhEBCHYA6RAHNdAchkJoFgLbse2PoLuP9AG4TYw1KSgowACMIHgxgHmgaKDgFRhwAYSFBTn4hAAUEX0S1E0w0BCkE9hloAH0zQRzqmULigC5UhiVsLKbcgUE2wUSAH2hBLhAQaABeCamQQNDocsrcmuho2PjEuuQctBVCTvpFJmPllcadrgSENANSCNNs3IAVBF1MIpHTtJ6fWaeSGvx4hlKJk03j4EGsDTeLWGiXmP2KyyhmjRI3hjWBNWRc2gMIgGM04SoyXk3SYABEwDwwECBgUwVopLJqedMoiQYTIRo1lUpLV9khDsd-jTlARCFdXoJbvdHmKJbKpdy8LKISsTBAuJh2I8grroT44aqjcciZajrKyWEqArgZ9vmCNelsPTGcyEoNkTrMRp9Ya0NY8R8vtjwQ7oBTqOAUmdPQBlPh2QRgEC+vlsiQcuRdTU5-1sgWrCrCza2yVFz1sTi8QaXJ2aG5xO48B66Gvqus9Bs6ZvaoxBtwGo0GWMmEnwg5W2U2+d20RylYvNu813Rj09NMZrMl1nFQNBkNG8NbqNgk0mePt6BRRfzHJIfsXWVT6jmRISRtSa0X3YN9kwHbQmyEY4v2-RVEhkBkmV5UtiiJV93xwb1EP6P1jxGXQn1XE0HwgdNM2zJDcJRCZgPQ7B9zIo9QWKfDtWCAB6NjoAAMzAPgQBgWYAD5oAQQgAEIYNiNxSMPCimJGAAmVCaNAph6Nk7Dc2Yv8dHYY4TQoeCfTk5FL0010jKw4FkJGIJoA48dgGAdgagE4TRMMhDGNMiMEldEiD3IzSbJjezOJ4vi3JEwgqAChiTLBMyXS+SzvJvdjwt4-i5ncmK4o06zKKSyNdHyoLCvkngFLshz9SclzCyE6KKCAA), that makes `_ownerBlock` abstract in the base class. It allows to define `_ownerBlock` with the right type in each subclass, and don't necessitate to declare it `| null`, so no need to use "!".

Reply:
TIL you can mark properties abstract 😮 (in retrospect, that makes a lot of sense since methods and properties are all the same thing anyway 😅 ) - this is awesome Alexis it organizes everything much more neatly 😄